### ISSUE TITLE
Support for Concurrent Forcemerge

### ISSUE THREAD

**Issue by danielkasen** [elastic/curator#1286]
So i'm trying to deal with a problem where I create 2 many indices a day to forcemerge using a single thread. I have a cluster dedicated to the FM process and it has multiple threads allocated to it specifically for this use case. The problem here is that curator only allows a single FM at a time and walks through the array of indices returned from the filter. It would be nice if I could toggle the level of concurrency so that I can run 2 FM at once for an index pattern without having to create a new action file. I know there are inherent risks with offering this and I could see a user not knowing to do this is most clusters, but I could really use it as I'm currently just abusing the timeout and continue_if_exception options.

**Comment by untergeek** [comment]
Unfortunately this is not a limitation of Curator, but of [Elasticsearch](
https://www.elastic.co/guide/en/elasticsearch/reference/6.4/indices-forcemerge.html#indices-forcemerge):

> This call will block until the merge is complete. If the http connection is lost, the request will continue in the background, and any new requests will block until the previous force merge is complete.

**Comment by danielkasen** [comment]
Right, but it won't block another call. For instance if you run a FM API call against 1 index then another 1 it'll put the second call into the FM queue or start processing it if you have increased the FM thread_pool:

GET _cat/thread_pool/force_merge?v&s=name

And here is a snippet of increasing the thread_pool in the elasticsearch.yml:

thread_pool:
  force_merge:
    size: "1"

**Comment by untergeek** [comment]
Now, Curator could be altered to permit parallel force merges, where 1 request submits multiple indices for force merge at once.  This is possible in the Elasticsearch API.  It's also extremely painful in terms of disk I/O for the nodes performing this task.  This is why Curator—as a matter of protecting the users from hammering their cluster too much—does the force merges serially.  All things considered, however, that approach is a can of worms.  You'd have to have dedicated "force merge" nodes in a shard-allocation routing setup, and ensure your shards are in there for that approach to work _safely._  Most users aren't going to have that, so I'd be wary of exposing them to a flag that permitted it.

**Comment by danielkasen** [comment]
So, i'm not sure if curator can fork here. Sry as I haven't dug into the code yet. But I would see it creating a child thread for each FM operation and letting us specify the number of child threads.

**Comment by untergeek** [comment]
Yeah, that approach is not safe to add to Curator.  I don't even approve of that approach in general.

**Comment by untergeek** [comment]
There are safety reasons why the developers chose what they did with force merge.  Just because a thing can be done doesn't mean it's best practices, or safe for general consumption.

**Comment by danielkasen** [comment]
haha yea, I figure i'm sort of an outlier here and totally understand protecting other users. As far as the approach goes, like I said i'm unfamiliar with your code so I'm just brainstorming here.

**Comment by untergeek** [comment]
And you're right, it doesn't "block" the call, but queues it.  Elasticsearch should not be spawning multiple parallel force merges simply because of the extreme amount of disk I/O it represents.  That's why it's clearly outlined that it will block subsequent force merges in the documentation.

**Comment by untergeek** [comment]
In the vein of “pics, or it didn’t happen,” have you actually observed concurrent forceMerge operations from successive calls without altering any settings from the defaults? At the index-level, not the shard level. And this should be one index per call, not a wildcard or csv indices.

**Comment by untergeek** [comment]
If you make a forceMerge call with wildcards, the concurrent merge throttling might have to be tweaked to permit more than 2 per node

**Comment by danielkasen** [comment]
Sure, so if I do
```
POST test-2018.09.30/_forcemerge?max_num_segments=5&flush=true
POST test-2018.09.29/_forcemerge?max_num_segments=5&flush=true
```
Then check the merge threads:
`GET _cat/thread_pool/force_merge?v&s=name`

I can see multiple running:
```
node_name           name        active queue rejected
node-1                  force_merge      2     0        0
node-2                  force_merge      2     0        0
node-3                  force_merge      2     0        0
node-4                 force_merge      2     0        0
```
And after I go past my thread_pool.force_merge.size of 2 it'll start adding them to the queue

**Comment by danielkasen** [comment]
Ahh yes I see, if I use a wildcard it'll only execute the FM operation to 1 thread, but it'll pick any shard that's available so this becomes more spread out in my cluster as all nodes are running an operation at once vs. only the nodes that have test-index-2018.09.30 it isn't adding all the other shards into the queue in this case so it doesn't take advantage of multiple thread_pools.

option 1) Allow wildcards in the call to allow all nodes to contribute at the same time
option 2) Allow the returned array by curator to be parallelized so that multiple POST operations happen at the same time.

**Comment by danielkasen** [comment]
Yup, on a node with 2 FM threads going:

```
{
  "nodes": {
    "GdJF3lveTeu8HGxXNv4zig": {
      "name": "XXXX",
      "transport_address": "XXXX",
      "host": "XXXX",
      "ip": "XXXX",
      "roles": [
        "data"
      ],
      "attributes": {
        "storage_type": "warm",
        "rack_id": "ap-southeast-1c",
        "instance_id": "XXXX",
        "xpack.installed": "true",
        "instance_type": "large",
        "instance_ip": "XXXX"
      },
      "tasks": {
        "GdJF3lveTeu8HGxXNv4zig:11959020": {
          "node": "GdJF3lveTeu8HGxXNv4zig",
          "id": 11959020,
          "type": "netty",
          "action": "indices:admin/forcemerge[n]",
          "start_time_in_millis": 1539283160172,
          "running_time_in_nanos": 4159561258352,
          "cancellable": false,
          "parent_task_id": "PKOs3zbfQwaFF9s5JkSWtQ:4517418",
          "headers": {}
        },
        "GdJF3lveTeu8HGxXNv4zig:11958857": {
          "node": "GdJF3lveTeu8HGxXNv4zig",
          "id": 11958857,
          "type": "netty",
          "action": "indices:admin/forcemerge[n]",
          "start_time_in_millis": 1539283118261,
          "running_time_in_nanos": 4201472806632,
          "cancellable": false,
          "parent_task_id": "8ozhh95EQlySvjsUX0sVPA:4415633",
          "headers": {}
        }
      }
    }
  }
}
```
